File: Tepora-app/scripts/build_sidecar.py
import os
import sys
import shutil
import subprocess
import zipfile
import tarfile
import re
from pathlib import Path

# Constants
PROJECT_ROOT = Path(__file__).resolve().parents[1]
BACKEND_DIR = PROJECT_ROOT / "backend"
FRONTEND_TAURI_DIR = PROJECT_ROOT / "frontend" / "src-tauri"
BINARIES_DIR = FRONTEND_TAURI_DIR / "binaries"
RESOURCES_DIR = FRONTEND_TAURI_DIR / "resources"
FALLBACK_DIR = RESOURCES_DIR / "llama-cpu-fallback"

# This should point to e:\Tepora_Project\格納
# PROJECT_ROOT is e:\Tepora_Project\Tepora-app
REPO_ROOT = PROJECT_ROOT.parent
STORAGE_DIR = REPO_ROOT / "格納"

# Determine current platform's target triple
# For Windows, usually x86_64-pc-windows-msvc
# You can run `rustc -vV` to get "host: ..." if rust is installed, or hardcode/detect.
def get_target_triple():
    # Simple detection for Windows x64
    if sys.platform == "win32":
        return "x86_64-pc-windows-msvc"
    elif sys.platform == "darwin":
        # Check architecture
        import platform
        if platform.machine() == "arm64":
            return "aarch64-apple-darwin"
        else:
            return "x86_64-apple-darwin"
    else:
        # Default to linux x64 for now
        return "x86_64-unknown-linux-gnu"

TARGET_TRIPLE = get_target_triple()
EXECUTABLE_NAME = "tepora-backend"
if sys.platform == "win32":
    FULL_EXECUTABLE_NAME = f"{EXECUTABLE_NAME}-{TARGET_TRIPLE}.exe"
else:
    FULL_EXECUTABLE_NAME = f"{EXECUTABLE_NAME}-{TARGET_TRIPLE}"

def clean():
    print("Cleaning previous builds...")
    dist_dir = BACKEND_DIR / "dist"
    build_dir = BACKEND_DIR / "build"
    spec_file = BACKEND_DIR / "server.spec"
    
    if dist_dir.exists(): shutil.rmtree(dist_dir)
    if build_dir.exists(): shutil.rmtree(build_dir)
    # Don't delete spec file if we want to reuse it, but here we might auto-generate.

    # Clean fallback resources
    if FALLBACK_DIR.exists():
        print(f"Cleaning fallback dir: {FALLBACK_DIR}")
        shutil.rmtree(FALLBACK_DIR)

def get_platform_variant_regex():
    """Get regex to match the correct fallback binary from '格納'"""
    # Filenames in 格納:
    # llama-b7574-bin-macos-arm64.tar.gz
    # llama-b7574-bin-ubuntu-x64.tar.gz
    # llama-b7574-bin-win-cpu-arm64.zip
    # llama-b7574-bin-win-cpu-x64.zip

    import platform
    machine = platform.machine().lower()

    if sys.platform == "win32":
        if machine in ("arm64", "aarch64"):
            return r"win-cpu-arm64\.zip$"
        else:
            return r"win-cpu-x64\.zip$"
    elif sys.platform == "darwin":
        if machine == "arm64":
            return r"macos-arm64\.tar\.gz$"
        else:
            return r"macos-x64\.tar\.gz$"
    else:
        # Linux
        return r"ubuntu-x64\.tar\.gz$" # Default to ubuntu-x64

def setup_fallback_binaries():
    print("Setting up fallback binaries...")
    if not STORAGE_DIR.exists():
        print(f"Warning: Storage directory not found at {STORAGE_DIR}. Skipping fallback setup.")
        return

    pattern = get_platform_variant_regex()
    regex = re.compile(pattern)

    found_archive = None
    for item in STORAGE_DIR.iterdir():
        if item.is_file() and regex.search(item.name):
            found_archive = item
            break
    
    if not found_archive:
        print(f"Warning: No matching fallback binary found in {STORAGE_DIR} for pattern {pattern}")
        return

    print(f"Found fallback archive: {found_archive}")
    FALLBACK_DIR.mkdir(parents=True, exist_ok=True)

    # Extract
    try:
        if found_archive.suffix == ".zip":
            with zipfile.ZipFile(found_archive, "r") as zf:
                zf.extractall(FALLBACK_DIR)
        elif found_archive.name.endswith(".tar.gz") or found_archive.name.endswith(".tgz"):
             with tarfile.open(found_archive, "r:gz") as tf:
                tf.extractall(FALLBACK_DIR)
        
        print(f"Extracted fallback binaries to {FALLBACK_DIR}")

        # Cleanup: Remove unneeded files if necessary (e.g. keeping only executables)
        # But commonly we need dlls too. Leaving as is is safer for now.
        
    except Exception as e:
        print(f"Error extracting fallback binary: {e}")
        sys.exit(1)


def build():
    # 1. Setup Resources
    setup_fallback_binaries()

    print(f"Building sidecar for {TARGET_TRIPLE}...")
    
    # Ensure binaries dir exists
    BINARIES_DIR.mkdir(parents=True, exist_ok=True)

    # Run PyInstaller
    # We use backend/server.py as entry point
    # We need to include src package
    # We might need to handle hidden imports or data files if config/logs usage requires it
    # BUT: Config is external (loaded from . or internal)
    
    cmd = [
        "pyinstaller",
        "--clean",
        "--noconfirm",
        "--name", EXECUTABLE_NAME,
        "--onefile",
        # "--windowed", # Uncomment if you don't want a console window on launch (debug useful though)
        "--distpath", str(BACKEND_DIR / "dist"),
        "--workpath", str(BACKEND_DIR / "build"),
        "--specpath", str(BACKEND_DIR),
        # Add paths
        "--paths", str(BACKEND_DIR),
        # Hidden imports often needed for uvicorn/fastapi/pydantic/langchain
        "--hidden-import", "uvicorn.logging",
        "--hidden-import", "uvicorn.loops",
        "--hidden-import", "uvicorn.loops.auto",
        "--hidden-import", "uvicorn.protocols",
        "--hidden-import", "uvicorn.protocols.http",
        "--hidden-import", "uvicorn.protocols.http.auto",
        "--hidden-import", "uvicorn.protocols.websockets",
        "--hidden-import", "uvicorn.protocols.websockets.auto",
        "--hidden-import", "uvicorn.lifespan.on",
        "--hidden-import", "sqlite3",
        "--hidden-import", "tiktoken_ext.openai_public",
        "--hidden-import", "tiktoken_ext",
        # Additional hidden imports for installer compatibility
        "--hidden-import", "huggingface_hub",
        "--hidden-import", "pydantic",
        "--hidden-import", "pydantic_core",
        "--hidden-import", "chromadb",
        "--hidden-import", "chromadb.config",
        "--hidden-import", "httpx",
        "--hidden-import", "httpcore",
        "--hidden-import", "psutil",
        # Collect submodules for complex packages
        # "--collect-submodules", "langchain", # Too heavy
        "--collect-submodules", "langchain_core",
        # "--collect-submodules", "langchain_community", # Too heavy, pulls torch
        "--collect-submodules", "chromadb",
        
        # Exact dependencies needed
        "--hidden-import", "langchain_openai",
        "--hidden-import", "langchain_text_splitters",
        
        # Excludes to reduce size
        "--exclude-module", "torch",
        "--exclude-module", "torchvision",
        "--exclude-module", "torchaudio",
        "--exclude-module", "transformers",
        "--exclude-module", "sentence_transformers",
        "--exclude-module", "nvidia",
        "--exclude-module", "triton",
        "--exclude-module", "sympy",
        # Entry point
        str(BACKEND_DIR / "server.py")
    ]
    
    print(f"Running: {' '.join(cmd)}")
    subprocess.check_call(cmd)

    # Move to src-tauri/binaries
    src_bin = BACKEND_DIR / "dist" / (f"{EXECUTABLE_NAME}.exe" if sys.platform == "win32" else EXECUTABLE_NAME)
    dst_bin = BINARIES_DIR / FULL_EXECUTABLE_NAME
    
    if src_bin.exists():
        print(f"Moving {src_bin} to {dst_bin}")
        shutil.move(str(src_bin), str(dst_bin))
        print("Build success!")
    else:
        print("Error: Build artifact not found.")
        sys.exit(1)

if __name__ == "__main__":
    clean()
    build()

File: Tepora-app/scripts/build_sidecar.mjs
import { execFileSync, spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PROJECT_ROOT = path.resolve(__dirname, "..");
const BACKEND_DIR = path.join(PROJECT_ROOT, "backend-rs");
const FRONTEND_TAURI_DIR = path.join(PROJECT_ROOT, "frontend", "src-tauri");
const BINARIES_DIR = path.join(FRONTEND_TAURI_DIR, "binaries");
const RESOURCES_DIR = path.join(FRONTEND_TAURI_DIR, "resources");
const FALLBACK_DIR = path.join(RESOURCES_DIR, "llama-cpu-fallback");

const REPO_ROOT = path.resolve(PROJECT_ROOT, "..");
const STORAGE_DIR = path.join(REPO_ROOT, "格納");

const EXECUTABLE_NAME = "tepora-backend";

function getTargetTriple() {
  const platform = process.platform;
  const arch = process.arch;

  if (platform === "win32") {
    if (arch === "arm64") {
      return "aarch64-pc-windows-msvc";
    }
    return "x86_64-pc-windows-msvc";
  }
  if (platform === "darwin") {
    if (arch === "arm64") {
      return "aarch64-apple-darwin";
    }
    return "x86_64-apple-darwin";
  }
  if (platform === "linux") {
    if (arch === "arm64") {
      return "aarch64-unknown-linux-gnu";
    }
    return "x86_64-unknown-linux-gnu";
  }

  throw new Error(`Unsupported platform ${platform} ${arch}`);
}

const TARGET_TRIPLE = getTargetTriple();
const FULL_EXECUTABLE_NAME =
  process.platform === "win32"
    ? `${EXECUTABLE_NAME}-${TARGET_TRIPLE}.exe`
    : `${EXECUTABLE_NAME}-${TARGET_TRIPLE}`;

function ensureDir(dir) {
  fs.mkdirSync(dir, { recursive: true });
}

function cleanFallbackDir() {
  if (fs.existsSync(FALLBACK_DIR)) {
    fs.rmSync(FALLBACK_DIR, { recursive: true, force: true });
  }
}

function getFallbackRegex() {
  if (process.platform === "win32") {
    return process.arch === "arm64"
      ? /win-cpu-arm64\.zip$/i
      : /win-cpu-x64\.zip$/i;
  }
  if (process.platform === "darwin") {
    return process.arch === "arm64"
      ? /macos-arm64\.tar\.gz$/i
      : /macos-x64\.tar\.gz$/i;
  }
  return /ubuntu-x64\.tar\.gz$/i;
}

function extractArchive(archivePath) {
  ensureDir(FALLBACK_DIR);

  if (archivePath.endsWith(".zip")) {
    if (process.platform === "win32") {
      execFileSync(
        "powershell",
        [
          "-NoProfile",
          "-Command",
          `Expand-Archive -Path \"${archivePath}\" -DestinationPath \"${FALLBACK_DIR}\" -Force`,
        ],
        { stdio: "inherit" }
      );
    } else {
      execFileSync("unzip", ["-o", archivePath, "-d", FALLBACK_DIR], {
        stdio: "inherit",
      });
    }
    return;
  }

  if (archivePath.endsWith(".tar.gz") || archivePath.endsWith(".tgz")) {
    execFileSync("tar", ["-xzf", archivePath, "-C", FALLBACK_DIR], {
      stdio: "inherit",
    });
    return;
  }

  throw new Error(`Unsupported archive format: ${archivePath}`);
}

function setupFallbackBinaries() {
  console.log("Setting up fallback binaries...");

  if (!fs.existsSync(STORAGE_DIR)) {
    console.warn(
      `Warning: Storage directory not found at ${STORAGE_DIR}. Skipping fallback setup.`
    );
    return;
  }

  const regex = getFallbackRegex();
  const entries = fs.readdirSync(STORAGE_DIR);
  const archive = entries.find((name) => regex.test(name));

  if (!archive) {
    console.warn(
      `Warning: No matching fallback binary found in ${STORAGE_DIR} for pattern ${regex}`
    );
    return;
  }

  const archivePath = path.join(STORAGE_DIR, archive);
  console.log(`Found fallback archive: ${archivePath}`);

  cleanFallbackDir();
  extractArchive(archivePath);
  console.log(`Extracted fallback binaries to ${FALLBACK_DIR}`);
}

function buildSidecar() {
  setupFallbackBinaries();

  console.log(`Building Rust sidecar for ${TARGET_TRIPLE}...`);
  ensureDir(BINARIES_DIR);

  const buildResult = spawnSync(
    "cargo",
    ["build", "--release", "--manifest-path", path.join(BACKEND_DIR, "Cargo.toml")],
    { stdio: "inherit" }
  );

  if (buildResult.status !== 0) {
    process.exit(buildResult.status ?? 1);
  }

  const backendBinary = path.join(
    BACKEND_DIR,
    "target",
    "release",
    process.platform === "win32" ? `${EXECUTABLE_NAME}.exe` : EXECUTABLE_NAME
  );

  if (!fs.existsSync(backendBinary)) {
    console.error(`Error: backend binary not found at ${backendBinary}`);
    process.exit(1);
  }

  const dstBinary = path.join(BINARIES_DIR, FULL_EXECUTABLE_NAME);
  fs.copyFileSync(backendBinary, dstBinary);
  console.log(`Copied ${backendBinary} to ${dstBinary}`);
  console.log("Build success!");
}

buildSidecar();
